week4/wordbreak.cc

```diff
@@ -0,0 +1,130 @@
+/* wordbreak.cc
+ * author: connie
+ *
+ * WORD BREAK
+ * Given an input string and a dictionary of words,
+ * find out if the input string can be segmented into a
+ * space-separated sequence of dictionary words.
+ *
+ */
+#include <iostream>
+#include <sstream>
+#include <cassert>
+#include <vector>
+#include <string>
+#include <map>
+
+using namespace std;
+
+//== DECLARATIONS
+bool wordInDict(string& word, vector<string>& dict);
+bool wordBreak_BF(string& s, vector<string>& dict);
+
+//== DEFINITIONS
+//#define DEBUG
+
+//== FUNCTIONS
+/*
+ * bool wordInDict(string& word, vector<string>& dict)
+ * inputs: word - word to check
+           dict - the dictionary (array of strings) with valid words
+ * outputs: True for valid dictionary word, False for invalid word
+ * notes: helper function for iterating thru the array
+ */
+bool wordInDict(string& word, vector<string>& dict)
+{
+  if (word.size() == 0) return false;
+
+  // iterate thru dictionary to check if the word is valid
+  for (string dictword : dict) {
+#ifdef DEBUG
+    cout << word << " ?= " << dictword << endl;
+#endif
+    if (word.compare(dictword) == 0) return true;
+  }
+
+  return false;
+}
+
+/*
+ * bool wordBreak_BF(string& s, vector<string>& dict)
+ * inputs: s - the string to check for a valid "word break"
+           dict - the dictionary (array of strings) with valid words
+ * outputs: True for valid word break, False for invalid word break
+ * notes: the brute force solution implementation - runs in O(n x m), where
+          n = size of string and m = size of dict
+ */
+bool wordBreak_BF(string& s, vector<string>& dict)
+{
+  // handle empty cases
+  if (s.size() == 0 || dict.size() == 0) return false;
+
+  // word breaks aren't case sensitive
+  for (int c = 0; c < s.size(); c++) {
+    s[c] = tolower(s[c]);
+  }
+
+  // helper vars for iterating thru s
+  string word = "";
+  int i = 0;
+
+  // iterate thru the string looking for words
+  while (i < s.size()) {
+    // build the word starting with each letter in the string
+    word += s[i];
+
+    // if word is valid dictionary word, trim s & reset the word
+    if (wordInDict(word, dict)) {
```

Comment:
oooh, thanks for catching that! the example given in the question totally threw me off this edge case. i fixed my function to now recursively check for the remainder of the string, to see if they are also in the dictionary as portions of compound words. :)